File: libraries/inventor2040w/inventor.cpp
#include <math.h>

#include "inventor.hpp"
#include "pwm.hpp"

namespace inventor {

Inventor2040W::Inventor2040W(float motor_gear_ratio)
: motors{Motor(MOTOR_A_PINS), Motor(MOTOR_B_PINS)}
, encoders{Encoder(pio0, 0, ENCODER_A_PINS, PIN_UNUSED, NORMAL_DIR, MMME_CPR * motor_gear_ratio, true),
           Encoder(pio0, 1, ENCODER_B_PINS, PIN_UNUSED, NORMAL_DIR, MMME_CPR * motor_gear_ratio, true)}
, servos{Servo(SERVO_1_PIN), Servo(SERVO_2_PIN), Servo(SERVO_3_PIN),
         Servo(SERVO_4_PIN), Servo(SERVO_5_PIN), Servo(SERVO_6_PIN)}
, leds(NUM_LEDS, pio0, 2, LED_DATA_PIN)
, i2c(I2C_SDA_PIN, I2C_SCL_PIN, 100000)
, audio_volume(DEFAULT_VOLUME) {
}

bool Inventor2040W::init(bool init_motors, bool init_servos) {
  bool success = true;

  // Set up the motors and encoders, if the user wants them
  if(init_motors) {
    for(uint i = 0; i < NUM_MOTORS && success; i++) {
      success = motors[i].init() && encoders[i].init();
    }
  }

  // Set up the servos, if the user wants them
  if(init_servos) {
    for(uint i = 0; i < NUM_SERVOS && success; i++) {
      success = servos[i].init();
    }
  }

  // Set up the amp enable
  gpio_set_function(AMP_EN_PIN, GPIO_FUNC_SIO);
  gpio_set_dir(AMP_EN_PIN, GPIO_OUT);
  gpio_put(AMP_EN_PIN, false);

  // Set up the audio pwm channel
  uint16_t period; uint16_t div16;
  if(pimoroni::calculate_pwm_factors(SILENCE_FREQUENCY, period, div16)) {
    pwm_config pwm_cfg = pwm_get_default_config();

    // Set the new wrap (should be 1 less than the period to get full 0 to 100%)
    pwm_config_set_wrap(&pwm_cfg, period - 1);

    // Apply the divider
    pwm_config_set_clkdiv(&pwm_cfg, (float)div16 / 16.0f); // There's no 'pwm_config_set_clkdiv_int_frac' for some reason...

    // Set up the phase and enable pins
    pwm_init(pwm_gpio_to_slice_num(PWM_AUDIO_PIN), &pwm_cfg, true);
    gpio_set_function(PWM_AUDIO_PIN, GPIO_FUNC_PWM);
    pwm_set_gpio_level(PWM_AUDIO_PIN, 0);

    success = true;
  }

  // Set up the user switch
  gpio_set_function(USER_SW_PIN, GPIO_FUNC_SIO);
  gpio_set_dir(USER_SW_PIN, GPIO_IN);
  gpio_pull_down(USER_SW_PIN);

  // Start the WS2812 LEDs
  leds.start();

  return success;
}

bool Inventor2040W::switch_pressed() {
  return gpio_get(USER_SW_PIN);
}

bool Inventor2040W::play_tone(float frequency) {
  bool success = false;

  // Calculate a suitable pwm wrap period for this frequency
  uint16_t period; uint16_t div16;
  if(pimoroni::calculate_pwm_factors(frequency, period, div16)) {

    uint8_t div = div16 >> 4;
    uint8_t mod = div16 % 16;

    uint pin_slice = pwm_gpio_to_slice_num(PWM_AUDIO_PIN);

    // Apply the new divider
    pwm_set_clkdiv_int_frac(pin_slice, div, mod);

    float corrected_volume = powf(audio_volume, 4.0f);
    uint16_t level = (uint16_t)((float)(period >> 1) * corrected_volume);

    // Set the new wrap (should be 1 less than the period to get full 0 to 100%)
    pwm_set_wrap(pin_slice, period - 1);

    // Update the pwm after setting the new wrap, and unmute
    pwm_set_gpio_level(PWM_AUDIO_PIN, level);
    unmute_audio();

    success = true;
  }
  else {
    play_silence();
  }
  return success;
}

void Inventor2040W::play_silence() {
  // Calculate a suitable pwm wrap period for the "silence" frequency
  uint16_t period; uint16_t div16;
  if(pimoroni::calculate_pwm_factors(SILENCE_FREQUENCY, period, div16)) {

    uint8_t div = div16 >> 4;
    uint8_t mod = div16 % 16;

    uint pin_slice = pwm_gpio_to_slice_num(PWM_AUDIO_PIN);

    // Apply the new divider
    pwm_set_clkdiv_int_frac(pin_slice, div, mod);

    float corrected_volume = powf(audio_volume, 4.0f);
    uint16_t level = (uint16_t)((float)(period >> 1) * corrected_volume);

    // Set the new wrap (should be 1 less than the period to get full 0 to 100%)
    pwm_set_wrap(pin_slice, period - 1);

    // Update the pwm after setting the new wrap, and unmute
    pwm_set_gpio_level(PWM_AUDIO_PIN, level);
    unmute_audio();
  }
}

void Inventor2040W::stop_playing() {
  pwm_set_gpio_level(PWM_AUDIO_PIN, 0);
  mute_audio();
}

float Inventor2040W::volume() const {
  return audio_volume;
}

void Inventor2040W::volume(float volume) {
  audio_volume = MIN(MAX(volume, 0.0f), 1.0f);
}

void Inventor2040W::mute_audio() {
  gpio_put(AMP_EN_PIN, false);
}

void Inventor2040W::unmute_audio() {
  gpio_put(AMP_EN_PIN, true);
}
}

File: micropython/modules_py/inventor.py
import gc
from machine import Pin, PWM
from pimoroni_i2c import PimoroniI2C
from plasma import WS2812
from motor import Motor
from servo import Servo
from encoder import Encoder, MMME_CPR

# IO Pin Constants
GP0 = 0
GP1 = 1
GP2 = 2

A0 = 26
A1 = 27
A2 = 28

GPIOS = (GP0, GP1, GP2, A0, A1, A2)
ADCS = (A0, A1, A2)


# Index Constants
MOTOR_A = 0
MOTOR_B = 1

SERVO_1 = 0
SERVO_2 = 1
SERVO_3 = 2
SERVO_4 = 3
SERVO_5 = 4
SERVO_6 = 5

LED_GP0 = 0
LED_GP1 = 1
LED_GP2 = 2
LED_A0 = 3
LED_A1 = 4
LED_A2 = 5
LED_SERVO_1 = 6
LED_SERVO_2 = 7
LED_SERVO_3 = 8
LED_SERVO_4 = 9
LED_SERVO_5 = 10
LED_SERVO_6 = 11


# Count Constants
NUM_GPIOS = 6
NUM_ADCS = 3
NUM_MOTORS = 2
NUM_SERVOS = 6
NUM_LEDS = 12


class Inventor2040W():
    AMP_EN_PIN = 3
    I2C_SDA_PIN = 4
    I2C_SCL_PIN = 5
    MOTOR_A_PINS = (6, 7)
    MOTOR_B_PINS = (8, 9)
    ENCODER_A_PINS = (19, 18)
    ENCODER_B_PINS = (17, 16)

    SERVO_1_PIN = 10
    SERVO_2_PIN = 11
    SERVO_3_PIN = 12
    SERVO_4_PIN = 13
    SERVO_5_PIN = 14
    SERVO_6_PIN = 15

    LED_DATA_PIN = 20
    PWM_AUDIO_PIN = 21
    USER_SW_PIN = 22

    AMP_CORRECTION = 4
    DEFAULT_VOLUME = 0.2

    def __init__(self, motor_gear_ratio=50, init_motors=True, init_servos=True):
        # Free up hardware resources
        gc.collect()

        # Set up the motors and encoders, if the user wants them
        self.motors = None
        if init_motors:
            cpr = MMME_CPR * motor_gear_ratio
            self.motors = [Motor(self.MOTOR_A_PINS), Motor(self.MOTOR_B_PINS)]
            # Set the encoders to use PIO 0 and State Machines 0 and 1
            self.encoders = [Encoder(0, 0, self.ENCODER_A_PINS, counts_per_rev=cpr, count_microsteps=True),
                             Encoder(0, 1, self.ENCODER_B_PINS, counts_per_rev=cpr, count_microsteps=True)]

        # Set up the servos, if the user wants them
        self.servos = None
        if init_servos:
            self.servos = [Servo(i) for i in range(self.SERVO_1_PIN, self.SERVO_6_PIN + 1)]

        # Set up the i2c for Qw/st and Breakout Garden
        self.i2c = PimoroniI2C(self.I2C_SDA_PIN, self.I2C_SCL_PIN, 100000)

        # Set up the amp enable
        self.__amp_en = Pin(self.AMP_EN_PIN, Pin.OUT)
        self.__amp_en.off()

        self.audio_pwm = PWM(Pin(self.PWM_AUDIO_PIN))
        self.__volume = self.DEFAULT_VOLUME

        # Set up the user switch
        self.__switch = Pin(self.USER_SW_PIN, Pin.IN, Pin.PULL_DOWN)

        # Set up the WS2812 LEDs, using PIO 0 and State Machine 2
        self.leds = WS2812(NUM_LEDS, 0, 2, self.LED_DATA_PIN)
        self.leds.start()

    def switch_pressed(self):
        return self.__switch.value()

    def play_tone(self, frequency):
        try:
            self.audio_pwm.freq(frequency)
        except ValueError:
            self.play_silence()
            raise ValueError("frequency of range. Expected greater than 0")

        corrected_volume = (self.__volume ** 4)  # Correct for RC Filter curve
        self.audio_pwm.duty_u16(int(32768 * corrected_volume))
        self.unmute_audio()

    def play_silence(self):
        self.audio_pwm.freq(44100)

        corrected_volume = (self.__volume ** 4)  # Correct for RC Filter curve
        self.audio_pwm.duty_u16(int(32768 * corrected_volume))
        self.unmute_audio()

    def stop_playing(self):
        self.audio_pwm.duty_u16(0)
        self.mute_audio()

    def volume(self, volume=None):
        if volume is None:
            return self.__volume

        if volume < 0.01 or volume > 1.0:
            raise ValueError("volume out of range. Expected 0.0 to 1.0")

        self.__volume = volume

    def mute_audio(self):
        self.__amp_en.off()

    def unmute_audio(self):
        self.__amp_en.on()
